Add unit weights

lbittarello opened the pull request:
Closes #135, #358 and #516

nalimilan commented on src/weights.jl:
```diff
@@ -250,6 +250,76 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
+values(wv::UnitWeights{S, T})    where {S, T} = fill(one(T), length(wv))
+isempty(wv::UnitWeights{S, T})   where {S, T} = iszero(wv.sum)
+length(wv::UnitWeights{S, T})    where {S, T} = Int(wv.sum)
+Base.size(wv::UnitWeights{S, T}) where {S, T} = Tuple(length(wv))
```
Import `size` for consistency.

lbittarello replied:
`size` was not imported, though we'd already extended it in line 27. I will explicitly import it in the next pull request, when we simplify the weight types.

nalimilan replied:
Better do this now, commits should make sense in isolation.

nalimilan commented on src/weights.jl:
```diff
@@ -250,6 +250,76 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
+All weight elements are identically one.
+""" UnitWeights
+
+eltype(wv::UnitWeights{S, T})    where {S, T} = T
```
Actually I suspect this isn't needed for `AbstractVector`. I could be removed for other types.

lbittarello replied:
We use it in `quantile`, but it's easy to replace it. I will drop it in the next pull request.

nalimilan replied:
What I mean is that there's a fallback in Base which gives the same result.

nalimilan commented on src/weights.jl:
```diff
@@ -250,6 +250,76 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
+
+eltype(wv::UnitWeights{S, T})    where {S, T} = T
+values(wv::UnitWeights{S, T})    where {S, T} = fill(one(T), length(wv))
+isempty(wv::UnitWeights{S, T})   where {S, T} = iszero(wv.sum)
```
No need for type parameters here now below.

nalimilan commented on src/weights.jl:
```diff
@@ -250,6 +250,76 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
+length(wv::UnitWeights{S, T})    where {S, T} = Int(wv.sum)
+Base.size(wv::UnitWeights{S, T}) where {S, T} = Tuple(length(wv))
+
+function Base.getindex(wv::UnitWeights{S, T}, i::Int) where {S, T}
```
Use  `Base.@propagate_inbounds` and put the bounds check inside `@boundscheck` (https://docs.julialang.org/en/latest/devdocs/boundscheck/). You can also write `i < 0 <= length(wv)`.

Same below.

nalimilan commented on src/weights.jl:
```diff
@@ -250,6 +250,76 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
 # NOTE: No variance correction is implemented for exponential weights
 
+struct UnitWeights{S<:Real, T<:Real} <: AbstractWeights{S, T, V where V<:Vector{T}}
+    sum::S
```
Sum is actually the length here, so maybe better call it `len`? Also, `S` can probably be replaced with just `Int` given that it represents the length of a vector.

lbittarello replied:
I replaced `sum` by `len` and added a method to the function `sum`.

nalimilan commented on src/weights.jl:
```diff
@@ -250,6 +250,76 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
+    (i > 0) & (i <= length(wv)) ? one(T) : Base.throw_boundserror(wv, i)
+end
+
+function Base.getindex(wv::UnitWeights{S, T}, i::AbstractRange{<:Int}) where {S, T}
```
This definition is redundant with the previous one. Also, maybe we need to accept `<:Integer`?

lbittarello replied:
Sorry, I didn't know that ranges were arrays till @Nosferican told me so.

nalimilan commented on src/weights.jl:
```diff
@@ -250,6 +250,76 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
+"""
+    varcorrection(w::UnitWeights, corrected=false)
+
+* `corrected=true`: ``\\frac{1}{\\sum w - \\sum {w^2} / \\sum w}``
```
The formula doesn't reflect the code. Though I would expect all three formulas for frequency, probability and analytical weights to give the same result for this special case, right? If so, probably worth mentioning.

lbittarello replied:
Sorry, mindless copying and pasting

nalimilan commented on src/weights.jl:
```diff
@@ -279,6 +352,17 @@ Base.sum(v::BitArray, w::AbstractWeights) = wsum(v, values(w))
 
+for v in (AbstractArray{<:Number}, BitArray, SparseArrays.SparseMatrixCSC, AbstractArray)
+    @eval begin
+        function Base.sum(v::$v, w::UnitWeights) where {S,T}
```
```suggestion
        function Base.sum(v::$v, w::UnitWeights)
```

nalimilan commented on src/weights.jl:
```diff
@@ -475,13 +559,22 @@ function wsum(A::AbstractArray{T}, w::AbstractVector{W}, dim::Int) where {T<:Num
     _wsum!(similar(A, wsumtype(T,W), Base.reduced_indices(axes(A), dim)), A, w, dim, true)
 end
 
+function wsum(A::AbstractArray{<:Number}, w::UnitWeights{S,T}, dim::Int) where {S,T}
```
No need for parameters:
```suggestion
function wsum(A::AbstractArray{<:Number}, w::UnitWeights, dim::Int)
```

Same elsewhere.

nalimilan commented on src/weights.jl:
```diff
@@ -475,13 +559,22 @@ function wsum(A::AbstractArray{T}, w::AbstractVector{W}, dim::Int) where {T<:Num
 
+function wsum(A::AbstractArray{<:Number}, w::UnitWeights{S,T}, dim::Int) where {S,T}
+    (size(A, dim) != length(w)) && throw(DimensionMismatch("Inconsistent array dimension."))
+    return sum(A, dims = dim)
```
No spaces around `=`. Same elsewhere.

nalimilan commented on src/weights.jl:
```diff
@@ -475,13 +559,22 @@ function wsum(A::AbstractArray{T}, w::AbstractVector{W}, dim::Int) where {T<:Num
 end
 
+function wsum(A::AbstractArray{<:Number}, w::UnitWeights{S,T}, dim::Int) where {S,T}
+    (size(A, dim) != length(w)) && throw(DimensionMismatch("Inconsistent array dimension."))
```
No need for parentheses:
```suggestion
    size(A, dim) != length(w) && throw(DimensionMismatch("Inconsistent array dimension."))
```
Same elsewhere.

lbittarello replied:
You know, I got errors if I didn't use parentheses in earlier versions of Julia, so I've just used them ever since. They don't seem necessary any longer though.

nalimilan replied:
They're still there. ;-)

lbittarello replied:
Zut. I tried to be more thorough this time.

nalimilan commented on src/weights.jl:
```diff
@@ -530,6 +623,15 @@ _mean(A::AbstractArray, w::AbstractWeights, dims::Nothing) =
 
+function StatsBase._mean(A::AbstractArray, w::UnitWeights{S,T}, dims::Nothing) where {S,T}
+    (length(A) != length(w)) && throw(DimensionMismatch("Inconsistent array dimension."))
+    return mean(A)
```
I suspect you can drop this method an just pass `dims=dims` even when it's `nothing`.

lbittarello replied:
So did I, but it doesn't actually work.

nalimilan replied:
Too bad. Though it would be shorter to use a single function, with a condition `dims === nothing ? mean(A) : mean(A, dims=dims)`.

lbittarello replied:
It isn't much shorter, because the methods have different boundary checks (which needed fixing, incidentally), so I left them as they are, but I am open to changing it.

nalimilan replied:
Ah, sorry.

nalimilan commented on src/weights.jl:
```diff
@@ -250,6 +250,76 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
+
+Base.getindex(wv::UnitWeights{S, T}, ::Colon) where {S, T} = fill(one(T), length(wv))
+
+"""
```
I'm not sure we should add this method since it's very different from others: it doesn't accept a vector as input. The constructor may be enough.

lbittarello replied:
I lean towards keeping it. One can reasonably expect colon indexes to work on arrays. It works for other weight types too (because they have a `values` subfield).

nalimilan replied:
Are we talking about the same thing? I don't see how colon indices are related. :-)

lbittarello replied:
Sorry: your comment came just after the definition of `getindex(wv::UnitWeights{S, T}, ::Colon)`, so I assumed that was what it referred to. I don't know which method you have in mind then.

nalimilan replied:
OK, let's keep it, but better use a default value for `T` here, and require people to use the constructor to choose a different type: I don't think it should be too common.

The default `T` could be either `Int` or `Bool`. The latter is a bit weird (it gives `true`), but it is sometimes used for a "strong zero" and it always preserves the type with multiplication: `Int8(1) * true === Int8(1)`, which isn't the case with `1`. Not sure whether that's really useful though, since weighted algorithms likely need to widen `Int8` at some point, and for floats `Int` and `Bool` behave the same:`1.0f0 * 1 === 1.0f0`. So maybe the simpler `Int` is better.

Nosferican replied:
```
x = uweights(6)
x[:]
type UnitWeights has no fields values
```
vs
```
x[:] == fill(one(1, 6))
```
I think that is good.

lbittarello replied:
The default constructor (`uweights(s)`) was already setting the element type to an integer type, since it used the type of `s`, but I made it more explicit in the latest commit.

nalimilan commented on src/weights.jl:
```diff
@@ -250,6 +250,65 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
+Base.size(wv::UnitWeights) = Tuple(length(wv))
+
+@propagate_inbounds function Base.getindex(wv::UnitWeights{S, T}, i::Integer) where {S, T}
+    @boundscheck checkbounds(wv,i)
```
```suggestion
    @boundscheck checkbounds(wv, i)
```

nalimilan commented on src/weights.jl:
```diff
@@ -250,6 +250,65 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
+end
+
+@propagate_inbounds function Base.getindex(wv::UnitWeights{S, T}, i::AbstractArray{<:Int}) where {S, T}
+    @boundscheck checkbounds(wv,i)
```
```suggestion
    @boundscheck checkbounds(wv, i)
```

nalimilan commented on src/weights.jl:
```diff
@@ -250,6 +250,65 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
+    one(T)
+end
+
+@propagate_inbounds function Base.getindex(wv::UnitWeights{S, T}, i::AbstractArray{<:Int}) where {S, T}
```
Actually I realize this definition isn't the most natural: we'd better return a new weights vector of the same type. But that's a separate issue since other types do the same thing.

nalimilan commented on docs/src/weights.md:
```diff
@@ -115,6 +115,15 @@ julia> eweights(t, r, 0.3)
 
+### `UnitWeights`
+
+Unit weights treat all observations equally.
```
I suggest this:
```suggestion
Unit weights are a special case in which all observations are given a weight equal to `1`. Using such weights is equivalent to computing unweighted statistics.

This type can notably be used when implementing an algorithm so that a only a weighted variant has to be written. The unweighted variant is then obtained by passing a `UnitWeights` object. This is very efficient since no weights vector is actually allocated.
```

Also, I'd put this before `Weights`, which is the fallback type with less methods defined (no `varcorrection`).

nalimilan commented on src/weights.jl:
```diff
@@ -250,6 +250,65 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
+
+* `corrected=true`: ``\\frac{n}{n - 1}``, where ``n`` is the length of the weight vector
+* `corrected=false`: ``\\frac{1}{n}``, where ``n`` is the length of the weight vector
+"""
```
Or maybe just "This definition is equivalent to the correction applied to unweighted data.".

nalimilan commented:
Can you also add tests?

lbittarello commented:
Tests added!

nalimilan commented on test/weights.jl:
```diff
@@ -447,6 +460,27 @@ end
+    @test_throws DimensionMismatch sum(a, wt, 4)
+    @test_throws DimensionMismatch mean(a, wt, dims=4)
+
+    @test quantile([1.0, 4.0, 6.0, 8.0, 10.0], uweights(5), [0.5]) ≈ [6.0]
```
```suggestion
    @test quantile([1.0, 4.0, 6.0, 8.0, 10.0], uweights(5), 0.5) ≈ 6.0
```

lbittarello replied:
I used an array here on purpose (because the next test doesn't), but I'm open to changing it if you prefer. We could actually only keep the test of `median`, since it calls `quantile(x, w, [0.5]) `.

nalimilan replied:
Probably better test both forms for both methods then. You never know where bugs can happen.

nalimilan commented on src/weights.jl:
```diff
@@ -250,6 +249,66 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
 # NOTE: No variance correction is implemented for exponential weights
 
+struct UnitWeights{S<:Integer, T<:Real} <: AbstractWeights{S, T, V where V<:Vector{T}}
+    len::S
```
As I noted, better hardcode this as `Int` for simplicity.

nalimilan replied:
Still need to address this comment.

lbittarello replied:
Does it look better now?

nalimilan commented:
Thanks. A few methods don't seem to be covered by tests: see dark red in the margin at https://codecov.io/gh/JuliaStats/StatsBase.jl/compare/416af5561d3efd11c78c82c269a9d7c7d26c681d...009b9d5b893bd9235dfd96895acfe1a3c1df81ea/diff.

lbittarello commented:
I added more tests.

As it turns out, we need `eltype` for unit weights or `show` throws an error (due to the absence of the `values` field), so I added it back.

nalimilan commented on src/weights.jl:
```diff
@@ -279,23 +280,28 @@ end
  1.0
  1.0
  1.0
+julia> uweights(Float64, 3)
```
Add empty line.

nalimilan commented on src/weights.jl:
````diff
@@ -250,6 +249,72 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
+# Examples
+```julia-repl
+julia> uweights(3)
+3-element UnitWeights{Int64,Float64}:
````
Types don't seem consistent with what the code does. I guess this explains the CI failure.

nalimilan commented on src/weights.jl:
````diff
@@ -250,6 +249,72 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
+ 1
+```
+"""
+uweights(::Type{T}, s::S) where {S, T} = UnitWeights{S, T}(s)
````
```suggestion
uweights(::Type{T}, s::S) where {S, T<:Real} = UnitWeights{S, T}(s)
```

nalimilan commented on src/weights.jl:
````diff
@@ -250,6 +249,72 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
+```
+"""
+uweights(::Type{T}, s::S) where {S, T} = UnitWeights{S, T}(s)
+uweights(s::S)            where S      = UnitWeights{Int, S}(s)
````
It doesn't really make sense to use the type of the length to choose the type of the elements. Better just use `Int`.

lbittarello replied:
Sorry, the parameters were in the wrong order

nalimilan commented:
> As it turns out, we need `eltype` for unit weights or `show` throws an error (due to the absence of the `values` field), so I added it back.

What if you also remove the `eltype` methods for `AbstractWeights`?

lbittarello commented:
You're right. I'd messed up my tests. I've amended the last commit to eliminate the reintroduction of `eltype` and include additional revisions.

nalimilan commented on test/weights.jl:
```diff
@@ -96,6 +96,23 @@ end
+    @test size(wv) === (3,)
+    @test sum(wv) === 3.
+    @test values(wv) == fill(1.0, 3)
+    @test varcorrection(wv) == 1/3
```
```suggestion
    @test Base.varcorrection(wv) == 1/3
```
But it would also be nice to test `var`, just in case.

lbittarello replied:
It seems that this change broke the tests... I don't know if `varcorrection` is defined in base at all.

nalimilan replied:
Sorry, I mean StatsBase.

nalimilan commented on src/weights.jl:
```diff
@@ -250,6 +249,72 @@ eweights(t::AbstractVector, r::AbstractRange, λ::Real) =
+Base.getindex(wv::UnitWeights{T}, ::Colon) where T = fill(one(T), length(wv))
+
+"""
+    uweights(::Type{T}, s::Integer) where T<:Real
```
Maybe put the one-arg method first?

nalimilan commented on src/weights.jl:
```diff
@@ -300,8 +300,8 @@ julia> uweights(3)
 """
-uweights(::Type{T}, s::Int) where {T<:Real} = UnitWeights{T}(s)
 uweights(s::Int)                            = UnitWeights{Int}(s)
+uweights(::Type{T}, s::Int) where {T<:Real} = UnitWeights{T}(s)
```
Also change order in the docstring?

oxinabox commented on src/weights.jl:
```diff
@@ -279,6 +347,17 @@ Base.sum(v::BitArray, w::AbstractWeights) = wsum(v, values(w))
+    @eval begin
+        function Base.sum(v::$v, w::UnitWeights)
+            if length(v) != length(w)
+                throw(DimensionMismatch("Inconsistent array dimension."))
```
Why does `UnitWeight` have length?
And if it has to have length then why do we need to actually check it?
Is this a useful feature vs it just always being considered the right size?
Like `LinearAlgebra.I`

lbittarello replied:
That's an interesting point. Here's an example. In computing the variance, we obtain the normalization factor through `varcorrection`. It depends on the number of observations, so we need the length of the weight vector. We should also ensure that the user has not passed the wrong length lest we use the wrong normalization factor. There are ways around it (for instance, a method for `variance` that bypasses the weights), but we want to help developers avoid defining separate methods. More generally, I think that boundary checks at different points help users catch mistakes before they snowball.

nalimilan replied:
Yes, it doesn't hurt to check lengths, and you can avoid it using `@inbounds`. I'd say that in general it's quite easy to get the appropriate length since it's the length of another input argument.

oxinabox replied:
Calculating the length isn't free for general iterators.

nalimilan replied:
Yeah, but in most cases we take arrays, right? Also you can always create an object of length `typemax` if you don't care about the checks.

oxinabox replied:
Mostly isn't always. Isn't `dropmissing` after all is one of the cases of iterators that are not free to take then length of?

These checks just don't feel great to me.
What error are we catching?
And how does that error relate to weighting?

If one is checking one got the amount of data that was expected, then isn't that better done with an explicit check on the data gathering code? So a good error can be thrown?
I feel like this just adds work for the user of weights, and the best case is they get a confusing error message.

I'm all for defensive programming, but that means checking the function you are defining can work. Not just checking arbitrary things in the calling code.


> Also you can always create an object of length typemax if you don't care about the checks.

That doesn't work, does it? This check is checking the lengths are equal, not checking a `>`.

nalimilan replied:
We don't support weights with `dropmissing` currently, the main reason being that in general we would need to skip weights corresponding to missing entries (instead of naively taking weights in their order of appearance).

> That doesn't work, does it? This check is checking the lengths are equal, not checking a `>`.

Right, it doesn't work here (I was thinking about custom functions which want to take a `UnitWeights` as default weights for simplicity of implementation). I guess we could allow something like `UnitWeights(nothing)` (or maybe `UnitWeights`) to create an object with a undefined length. But I still think it's nice to take a length by default, for checking, but also because that's required to be an actual `AbstractVector` (otherwise we need to error when calling `size`).

lbittarello commented:
Happy to! It'll take me a couple of weeks though, as I'll be out of town.

nalimilan commented:
Gentle bump about `value`: it would be nice to have that before we make a release to avoid changing (slightly) the API.

matthieugomez commented:
I know I am a bit late but why not import `FillArrays` and  define 
`UnitWeights{T} = AbstractWeights{Int, T, Ones{T}}`.

This would simplify the code. It would also make it more efficient. For instance,  `sqrt.(uweights(10)` creates a full vector in this pull request, whereas `sqrt.(Ones{Float64}(10))` creates a `FillArray`.

nalimilan commented:
The issue with that approach is that `UnitWeights` wouldn't be a concrete type, and objects created by `uweights` wouldn't be `<:AbstractWeights`.
